#' @title RunARRIBA
#' @description Executes the Arriba software from the bam file generated by STAR.
#' @param patient_dir Path of the directory that contains bam file generated by STAR
#' @export

runARRIBA <- function(patient_dir, genomeversion = "hg38", assemblyVersion = "GRCh38") {

  ARRIBA <- downloadArriba(soft_directory)
  DB_Arriba <- sprintf("%s/database", dirname(ARRIBA))

  STAR <- downloadSTAR(soft_directory)

  out<- downloadHG38(soft_directory)
  AnnotationHG38 <- out[[2]]
  FastaHG38 <- out[[1]]
  #index_dir <- out[[3]]

  file_list <- list.files(patient_dir)

  if  (basename(patient_dir) == "trimmed") {
    patient_id <- basename(dirname(patient_dir))
  } else {
    patient_id <- basename(patient_dir)
  }

  bam_file <- paste0(patient_dir, "/", file_list[endsWith(file_list, "Aligned_out.bam")], sep="")
  fusion.file <- sprintf("%s/%s_fusions.tsv", patient_dir, patient_id)
  fusion.discarded.file <- sprintf("%s/%s_fusions_discarded.tsv", patient_dir, patient_id)

  #Evita repetir el analisis si ya fue hecho
  if (!(length(nchar(file_list[endsWith(file_list, "FusionReport.xlsx")])) == 0)) {
    message("The ARRIBA for this sample has already been done.")
    #return(paste0(patient_dir, "/", file_list[endsWith(file_list, "FusionReport.xlsx")], sep=""))

  } else {
    if ((length(nchar(bam_file)) == 0)) {
      stop("There are no BAM files in this directory")
    }

    system2(command = ARRIBA,
            args = c(paste0("-x ", bam_file),
                     #"-I",
                     paste0("-o ", fusion.file),
                     paste0("-O ", fusion.discarded.file),
                     paste0("-a ", FastaHG38),
                     paste0("-g ", AnnotationHG38),
                     paste0("-b ", sprintf("%s/blacklist_%s_%s_v2.4.0.tsv.gz", DB_Arriba, genomeversion, assemblyVersion)),
                     paste0("-k ", sprintf("%s/known_fusions_%s_%s_v2.4.0.tsv.gz", DB_Arriba, genomeversion, assemblyVersion)),
                     paste0("-t ", sprintf("%s/known_fusions_%s_%s_v2.4.0.tsv.gz", DB_Arriba, genomeversion, assemblyVersion)),
                     paste0("-p ", sprintf("%s/protein_domains_%s_%s_v2.4.0.gff3", DB_Arriba, genomeversion,assemblyVersion))
            ))

  }

  #if(!all(file.exists(fusion.file,fusion.discarded.file))){
  #  cat(paste0("\nFUSION FILES NOT FOUND\n",fusion.file,"\n", fusion.discarded.file))
  #}

  fusionsTable <- generateFusionsReport(fusion.file, assemblyVersion, patient_id, patient_dir)
  message("ARRIBA's analysis has finished!")

  # DRAW FUSIONS: -------------------------------------------------------------------------------
  sorted_bam <- paste0(patient_dir, "/", patient_id, "_sorted.bam", sep="")
  Samtools <- downloadSamtools(soft_directory)
  
  if(!file.exists(sorted_bam)) {

    # Sort BAM
    system2(Samtools, args = c("sort", "-@","8", "-o", sorted_bam, bam_file))

    # Index sorted BAM
    system2(Samtools, args = c("index", "-@","8", sorted_bam))

  }

  #AnnotationHG38="/home/juan/R/x86_64-pc-linux-gnu-library/4.1/OMICsdoSof/HG38/Homo_sapiens.GRCh38.110.gtf"
  fusions_pdf <- sprintf("%s/%s_fusions.pdf", patient_dir, patient_id)

  if(!file.exists(fusions_pdf)) {
    message("The fusions will be drawn up")
    system2(command = sprintf("%s/draw_fusions.R", dirname(ARRIBA)),
          args = c(paste0("--fusions=", fusion.file),
                   paste0("--alignments=", sorted_bam),
                   paste0("--output=", fusions_pdf),
                   paste0("--annotation=", AnnotationHG38),
                   paste0("--cytobands=", DB_Arriba, "/cytobands_hg38_GRCh38_v2.4.0.tsv" ),
                   paste0("--proteinDomains=", DB_Arriba, "/protein_domains_hg38_GRCh38_v2.4.0.gff3" )
          ))

  }

  message("The fusions have been drawn")
  return(fusionsTable)
}

#' @title generate Fusions Report
#' @description Generates an xlsx file with the information of the fusions detected by ARRIBA.
#' @param fusion.file Path of the fusions.tsv file generated by ARRIBA.
#' @param assemblyVersion version of the genome reference
#' @param patient_id string with patient identification
#' @param patient_dir path of the patient's folder.
#' @import openxlsx
generateFusionsReport <- function(fusion.file, assemblyVersion, patient_id, patient_dir) {

  fusionsTable <- ReadArribaFusionTable(fusion.file)
  version.info <- data.frame(  STAR = basename(dirname(dirname(dirname(STAR)))),
                               ARRIBA = basename(dirname(ARRIBA)),
                               Version = assemblyVersion,
                               SAMPLE = patient_id )
  #openxlsx::write.xlsx(list(Fusions=fusionsTable,Info=version.info), stringr::str_replace(fusion.file,".tsv",".xlsx"))


  wb <- openxlsx::createWorkbook()

  COLNAMES_STYLE <- createStyle(
    fontSize = 12,
    textDecoration = "bold",
    halign = "center", valign = "center", border = "TopBottom",
    borderColour = "black")

  addWorksheet(wb, paste(patient_id, "_Fusions"))
  addWorksheet(wb, "Info")
  writeData(wb, sheet = 1, x = fusionsTable,
            headerStyle = COLNAMES_STYLE,
            borderStyle = "dashed",
            borders = "columns", borderColour = "black")

  setColWidths(wb, sheet = 1, cols = 1:ncol(fusionsTable), widths = "auto")
  writeData(wb, sheet = 2, x = version.info ,
            headerStyle = COLNAMES_STYLE,
            borderStyle = "dashed",
            borders = "columns", borderColour = "black")

  openxlsx::saveWorkbook(wb, paste(patient_dir, "/", patient_id, "_FusionReport", ".xlsx", sep = ""),  overwrite = TRUE)
  message(sprintf("A report has been created with the information of the fusions detected. It is located in %s", paste(patient_dir, "/", patient_id, "_FusionReport", ".xlsx", sep = "")) )
  return(fusionsTable)
}




#' .ReadArribaFustionTable
#' Read and format arriba output fusion table
#' @param fusionFile string with the full path of the Fussion.tsv table
#' @import stringr
ReadArribaFusionTable <- function(fusionsFile ) {

  if(stringr::str_detect(fusionsFile,".xlsx")){
    fusions <- openxlsx::read.xlsx(fusionsFile)
    first.pattern <- "gene1"
    colnames(fusions) <- make.names(colnames(fusions))
    attr(fusions,"AssemblyInfo") <- openxlsx::read.xlsx(fusionsFile,sheet = "Info")
  } else {
    fusions <- read.table(fusionsFile, stringsAsFactors=F, sep="\t", header=T, comment.char="", quote="")
    first.pattern <- "X.gene1"
    attr(fusions,"AssemblyInfo") <- "Unknown"
  }
  # fusions <- read.table(fusionsFile, stringsAsFactors=F, sep="\t", header=T, comment.char="", quote="")

  if (colnames(fusions)[1] == first.pattern) { # Arriba output
    colnames(fusions)[colnames(fusions) == first.pattern] <- "gene1"
    colnames(fusions)[colnames(fusions) == "strand1.gene.fusion."] <- "strand1"
    colnames(fusions)[colnames(fusions) == "strand2.gene.fusion."] <- "strand2"

    if(first.pattern == "X.gene1") {
      fusions$display_contig1 <- sub(":[^:]*$", "", fusions$breakpoint1, perl=T)
      fusions$display_contig2 <- sub(":[^:]*$", "", fusions$breakpoint2, perl=T)
    }
    fusions$contig1 <- removeChr(fusions$display_contig1)
    fusions$contig2 <- removeChr(fusions$display_contig2)
    fusions$breakpoint1 <- as.numeric(sub(".*:", "", fusions$breakpoint1, perl=T))
    fusions$breakpoint2 <- as.numeric(sub(".*:", "", fusions$breakpoint2, perl=T))
    fusions$split_reads <-  as.numeric(fusions$split_reads1) + as.numeric(fusions$split_reads2)

    # fusions$contig1 <- removeChr(fusions$display_contig1)
    # fusions$contig2 <- removeChr(fusions$display_contig2)

  } else if (colnames(fusions)[1] == "X.FusionName") { # STAR-Fusion
    fusions$gene1 <- sub("\\^.*", "", fusions$LeftGene, perl=T)
    fusions$gene2 <- sub("\\^.*", "", fusions$RightGene, perl=T)
    # fusions$contig1 <- removeChr(fusions$display_contig1)
    # fusions$contig2 <- removeChr(fusions$display_contig2)
    fusions$strand1 <- sub(".*:(.)$", "\\1/\\1", fusions$LeftBreakpoint, perl=T)
    fusions$strand2 <- sub(".*:(.)$", "\\1/\\1", fusions$RightBreakpoint, perl=T)
    fusions$display_contig1 <- sub(":[^:]*:[^:]*$", "", fusions$LeftBreakpoint, perl=T)
    fusions$display_contig2 <- sub(":[^:]*:[^:]*$", "", fusions$RightBreakpoint, perl=T)
    fusions$contig1 <- removeChr(fusions$display_contig1)
    fusions$contig2 <- removeChr(fusions$display_contig2)
    fusions$breakpoint1 <- as.numeric(sub(".*:([^:]*):[^:]*$", "\\1", fusions$LeftBreakpoint, perl=T))
    fusions$breakpoint2 <- as.numeric(sub(".*:([^:]*):[^:]*$", "\\1", fusions$RightBreakpoint, perl=T))
    fusions$direction1 <- ifelse(grepl(":\\+$", fusions$LeftBreakpoint, perl=T), "downstream", "upstream")
    fusions$direction2 <- ifelse(grepl(":\\+$", fusions$RightBreakpoint, perl=T), "upstream", "downstream")
    fusions$transcript_id1 <- ifelse(rep(!("CDS_LEFT_ID" %in% colnames(fusions)), nrow(fusions)), ".", fusions$CDS_LEFT_ID)
    fusions$transcript_id2 <- ifelse(rep(!("CDS_RIGHT_ID" %in% colnames(fusions)), nrow(fusions)), ".", fusions$CDS_RIGHT_ID)
    fusions$fusion_transcript <- ifelse(rep(!("FUSION_CDS" %in% colnames(fusions)), nrow(fusions)), ".", toupper(sub("([a-z]*)", "\\1|", fusions$FUSION_CDS)))
    fusions$reading_frame <- ifelse(rep(!("PROT_FUSION_TYPE" %in% colnames(fusions)), nrow(fusions)), ".", ifelse(fusions$PROT_FUSION_TYPE == "INFRAME", "in-frame", ifelse(fusions$PROT_FUSION_TYPE == "FRAMESHIFT", "out-of-frame", ".")))
    fusions$split_reads <- fusions$JunctionReadCount
    fusions$discordant_mates <- fusions$SpanningFragCount
    fusions$site1 <- rep("exon", nrow(fusions))
    fusions$site2 <- rep("exon", nrow(fusions))
    fusions$confidence <- rep("high", nrow(fusions))

  } else {
    stop("Unrecognized fusion file format")
  }

  return(invisible(fusions))
}

removeChr <- function(contig) {
  sub("^chr", "", sub("^chrM", "MT", contig), perl=T)
}
