#' @title RunSTAR
#' @description Executes the alignment of the sample with STAR software.
#' @param patient_dir Path of the directory that contains R1 and R2 fastq files. It can be either the "trimmed" folder or the original folder.
#' @param twoPass kjvdfk
#' @param nThreads vlkdnf
#' @return path of the BAM generated by STAR
#' @export
runSTAR <- function(patient_dir, twoPass = c("None","Basic"), soft_directory) {

  #STAR <- downloadSTAR()

  out <- downloadHG38(soft_directory)
  AnnotationHG38 <- out[[2]]
  FastaHG38 <- out[[1]]
  index_dir <- out[[3]]

  twoPass <- match.arg(twoPass[1], choices = c("None","Basic"))

  #if(missing(nThreads)){
    nThreads <- max(1,parallel::detectCores()-2)
  #}

  if(nThreads < 4){
    message(paste0("STAR running on ",nThreads, " may be not optimal"))
  }else{
    message(paste0("STAR running on ",nThreads))
  }


  file_list <- list.files(patient_dir)
  id <- basename(patient_dir)
  if(id  == "trimmed") {
    id <- basename(dirname(patient_dir))
  }

  bam_file <- sprintf("%s/trimmed/%s_Aligned_out.bam", patient_dir, id)
  #Evita repetir el analisis si ya fue hecho
  if(file.exists(bam_file)) {
    if(file.info(bam_file)$size > 0) {
      message("The STAR alignment for this sample has already been done.")
      return(bam_file)
    }
  }

  #Para que se pueda poner como entrada la carpeta de los trimmeados o la carpeta original:
  if  (basename(patient_dir) == "trimmed") {
    fileR1 <- paste0(patient_dir, "/", file_list[endsWith(file_list, "val_1.fq.gz")], sep="")
    fileR2 <- paste0(patient_dir, "/", file_list[endsWith(file_list, "val_2.fq.gz")], sep="")
    patient_id <- basename(dirname(dirname(fileR1)))

  } else {
    fileR1 <- paste0(patient_dir, "/", file_list[endsWith(file_list, "R1.fastq.gz")], sep="")
    fileR2 <- paste0(patient_dir, "/", file_list[endsWith(file_list, "R2.fastq.gz")], sep="")
    patient_id <- basename(dirname(fileR1))
  }

  if ((length(nchar(fileR1)) == 0) | (length(nchar(fileR2)) == 0)) {
    stop("There are no fastq files in this directory")
  }

  setwd(patient_dir)
  log <- system2(command = STAR,
                 args = c(paste0("--runThreadN ", nThreads),
                          paste0("--genomeDir " , index_dir),
                          paste0("--readFilesIn ", fileR1, " ", fileR2),
                          paste0("--readFilesCommand ", ifelse(stringr::str_detect(fileR1,".gz"),"zcat","-")),
                          # "--outStd BAM_Unsorted",
                          "--outSAMtype BAM Unsorted",
                          "--outSAMunmapped Within",
                          "--outReadsUnmapped Fastx", # THIS WAS ADDED TO KEEP UNMAPPED READS IN SEPARATE .MATE FILES
                          "--outBAMcompression 0",
                          "--outFilterMultimapNmax 50",
                          "--peOverlapNbasesMin 10",
                          "--alignSplicedMateMapLminOverLmate 0.5",
                          "--alignSJstitchMismatchNmax 5 -1 5 5",
                          "--chimSegmentMin 10",
                          "--chimOutType WithinBAM HardClip",
                          "--chimJunctionOverhangMin 10",
                          "--chimScoreDropMax 30",
                          "--chimScoreJunctionNonGTAG 0",
                          "--chimScoreSeparation 1",
                          "--chimSegmentReadGapMax 3",
                          "--chimMultimapNmax 50",
                          #sprintf("--outTmpDir %s/trimmed/%s_STARgenome", patient_dir, id ),
                          paste0("--outFileNamePrefix ", patient_id),
                          paste0("--twopassMode ", twoPass),
                          paste0("--sjdbGTFfile ", AnnotationHG38)
                 ), stdout = TRUE)#out.file)

  #Rename de bam file
  file.rename(sprintf("%s/%sAligned.out.bam", patient_dir, patient_id), sprintf("%s/%s_Aligned_out.bam", patient_dir, patient_id))

  message("STAR's analysis has finished!")

  return(sprintf("%s/%s_Aligned_out.bam", patient_dir, patient_id))
}
